### ISSUE TITLE
Active mode makes ftp.py crash

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3596]
|<img alt="sylvinus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| sylvinus reported|
|-|-|
|Trac ID|trac#3596|
|Type|defect|
|Created|2008-12-29 22:27:06Z|
|Branch|https://github.com/twisted/twisted/tree/dtpfactory-timeouts-3596|

The following commands make a the twisted example ftp server crash :

```
--------------------------------------------------------------------------
WinSCP Version 4.1.8 (Build 415) (OS 5.1.2600 Service Pack 3)
Login time: lundi 29 décembre 2008 23:22:59
--------------------------------------------------------------------------
Session name: xxx@127.0.0.1
Host name: 127.0.0.1 (Port: 8021)
User name: xxx (Password: Yes, Key file: No)
Tunnel: No
Transfer Protocol: FTP
Ping type: C, Ping interval: 30 sec; Timeout: 15 sec
Proxy: none
FTP: Passive: No
Local directory: default, Remote directory: home, Update: No, Cache: Yes
Cache directory changes: Yes, Permanent: Yes
DST mode: 1
--------------------------------------------------------------------------
Connecting to 127.0.0.1:8021 ...
Connected with 127.0.0.1:8021. Waiting for welcome message...
220 Jamendo Partner Delivery FTP
USER xxx
331 Password required for xxx.
PASS ******************
230 User logged in, proceed
SYST
215 UNIX Type: L8
FEAT
502 Command 'FEAT' not implemented
Connected
--------------------------------------------------------------------------
Using FTP protocol.
Doing startup conversation with host.
PWD
257 "/"
Getting current directory name.
Retrieving directory listing...
TYPE A
200 Type set to A.
PORT 172,16,83,128,5,102
550 Requested action not taken: internal server error
```

You get this in the logs:

```
Traceback (most recent call last):
Failure: twisted.internet.defer.TimeoutError: 
Traceback (most recent call last):
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/internet/tcp.py", line 362, in doRead
    return self.protocol.dataReceived(data)
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/protocols/policies.py", line 72, in dataReceived
    self.wrappedProtocol.dataReceived(data)
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/protocols/basic.py", line 231, in dataReceived
    why = self.lineReceived(line)
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/protocols/ftp.py", line 698, in lineReceived
    d = defer.maybeDeferred(self.processCommand, cmd, *args)
--- <exception caught here> ---
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/internet/defer.py", line 106, in maybeDeferred
    result = f(*args, **kw)
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/protocols/ftp.py", line 729, in processCommand
    return method(*params)
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/protocols/ftp.py", line 844, in ftp_PORT
    self.cleanupDTP()
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/protocols/ftp.py", line 1220, in cleanupDTP
    dtpPort.disconnect()
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/internet/base.py", line 892, in disconnect
    self.stopConnecting()
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/internet/base.py", line 916, in stopConnecting
    self.transport.failIfNotConnected(error.UserError())
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/internet/tcp.py", line 479, in failIfNotConnected
    self.connector.connectionFailed(failure.Failure(err))
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/internet/base.py", line 936, in connectionFailed
    self.factory.clientConnectionFailed(self, reason)
  File "/home/neoexport/exports/jamendo_superjamseeder_beta8_1230048960-23929/twisted/protocols/ftp.py", line 547, in clientConnectionFailed
    self.deferred.errback(PortConnectionError(reason))
exceptions.AttributeError: DTPFactory instance has no attribute 'deferred'
```

Attachments:

* [active_ftp.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8db/8db948b772faeb9a1009081127669e12b62db2fe/4656a7ff956d41b677e35ad1bb6338b3e1c1df83.py) (4244 bytes) - added by sylvinus on 2009-01-08 08:42:19Z - unit test for the error

<details><summary>Searchable metadata</summary>

```
trac-id__3596 3596
type__defect defect
reporter__sylvinus sylvinus
priority__normal normal
milestone__ 
branch__branches_dtpfactory_timeouts_3596 branches/dtpfactory-timeouts-3596
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1230589626000000 1230589626000000
changetime__1231525139000000 1231525139000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

I can't easily reproduce the reported problem.  If I issue a PORT command, then it succeeds:

```
PORT 127,0,0,1,48,57
200 PORT OK
```

with no errors in the log.  Can you provide more details about how to reproduce this behavior?

**Comment by twisted-trac** [comment]
|<img alt="sylvinus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|sylvinus commented|
|-|-|

I attached a unit test for the error. It happens always when the IP sent via "PORT" is not accepting connections.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [25903]) Branching to 'dtpfactory-timeouts-3596'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

Fixed handling of timeout/connection failure/connection success interaction wrt DTPFactory in the branch.  Build results are at http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/dtpfactory-timeouts-3596

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

Great changes, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [25928]) Merge dtpfactory-timeouts-3596

Author: exarkun
Reviewer: therve
Fixes: [#3596](https://github.com/twisted/twisted/issues/3596)

Fix the interaction between DTP connection setup success, timeout, and failure
events.  In particular, fix the behavior when a connection fails after the
specified timeout has elapsed.  Add direct unit tests for DTPFactory as well.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
